Upload video caption

ikovac opened the pull request:
This PR introduces video captions.

**Demo:**
![Kapture 2021-02-04 at 11 24 59](https://user-images.githubusercontent.com/26693037/106879620-ac542a00-66db-11eb-8675-7d5b44c17e83.gif)

ikovac committed:
Add confirmation modal component

ikovac committed:
Add caption class and deleting state

ikovac committed:
Refactor initialState

ikovac committed:
Add upload caption option inside toolbar

ikovac committed:
Add beforeSave hook and caption upload logic

ikovac committed:
Rebuild dist

ikovac committed:
Delete caption prop before saving

ikovac committed:
Add global error message

ikovac committed:
Rename caption upload btn isDisabled computed prop

ikovac committed:
Refactor FileReader load event

ikovac committed:
Add error serialization

ikovac committed:
Fix height bug when a global error is displayed

ikovac committed:
Move default error to shared folder

ikovac committed:
🎨

ikovac committed:
Rename delete caption modal btn

ikovac committed:
Clean ConfirmationDialog 🎨

ikovac committed:
Rename btn isDisabled computed prop

ikovac committed:
🎨

ikovac committed:
🎨  Refactor beforeSave hook

ikovac committed:
Merge branch 'feature/implement-sprout-video' into feature/captions

ikovac committed:
Remove unnecessary caption error handling

ikovac committed:
Merge branch 'feature/implement-sprout-video' into feature/captions

ikovac committed:
Fix bugs after merging

ikovac committed:
Merge branch 'feature/implement-sprout-video' into feature/captions

ikovac committed:
🎨  Fix prop order

ikovac commented on src/edit/ConfirmationDialog.vue:
```diff
@@ -0,0 +1,36 @@
+<template>
+  <v-dialog v-model="dialog" width="500">
```
This is replaced with TailorDialog inside: https://github.com/ExtensionEngine/tce-sprout-video/pull/3/files#diff-fd06d3d1b05de79c059e83f698f712fe09fd734f38775a9de1d261ea1a8766f0R1

ikovac commented on src/edit/Toolbar.vue:
```diff
@@ -1,24 +1,53 @@
-    height="72"
-    color="transparent"
-    class="elevation-0">
+  <v-toolbar height="72" color="transparent" class="elevation-0">
```
The Toolbar component is also much simplified with https://github.com/ExtensionEngine/tce-sprout-video/pull/3/files#diff-c31caae4002cde1610e57406724be0233346402b6ab29f403e6da13536104c6bR2

ikovac committed:
Remove unnecessary style

MiroDojkic commented on src/edit/Toolbar.vue:
```diff
@@ -1,24 +1,53 @@
+            </v-icon>
+          </confirmation-dialog>
+        </template>
+      </v-text-field>
```
Should we move this to i.e. `CaptionField` to keep this component simple to read?

MiroDojkic replied:
Same question for video upload field. 😄

ikovac committed:
Implement TailorDialog inside ConfirmationDialog

ikovac committed:
🚀

MiroDojkic commented on src/edit/Toolbar.vue:
```diff
@@ -28,15 +57,49 @@ export default {
+    isCaptionUploadDisabled() {
+      const { id: videoId, playable } = this.element.data.video;
+      return !videoId || !playable || this.captionFileName;
+    }
```
Would it be more easier to read from the template if we'd replace this with `hasPlayableVideo` computed (or some better name 😅) that returns `videoId && playable` and then calculate `disable` prop explicitly:
```html
<upload-btn :disabled="!hasPlayableVideo || this.captionFileName" />
```

ikovac committed:
Add VideoUpload toolbar component

MiroDojkic commented on src/edit/index.vue:
```diff
@@ -38,73 +45,77 @@ export default {
-        status: ELEMENT_STATE.UPLOADING
-      });
+    this.$elementBus.on('save', ({ video, caption }) => {
+      if (video) {
+        this.file = video.file;
+        delete video.file;
+      }
```
Why do we need to do this?

ikovac replied:
Because we don't want the `file` prop to be part of the asset.data object which is sent to the server on the `save` event.

ikovac replied:
I manage to make it more readable using lodash methods.
```js
this.file = get(video, 'file', null);
const data = cloneDeep(this.element.data);
Object.assign(data.video, omit(video, ['file']));
Object.assign(data.caption, caption);
this.$emit('save', data);
```

MiroDojkic commented on src/server/index.js:
```diff
@@ -4,47 +4,84 @@ const { createClient } = require('./sproutVideo');
+  const { sproutVideoApiKey: apiKey } = tce;
+  const client = createClient({ apiKey });
+  const { id: videoId, playable } = asset.data.video;
+  const isVideoPlayable = videoId && playable;
```
Not sure if we need to alias this condition, seems self-explanatory. 🤔

MiroDojkic commented on src/server/index.js:
```diff
@@ -4,47 +4,84 @@ const { createClient } = require('./sproutVideo');
+    video: { id: videoId },
+    caption: { id: captionId, content, status }
+  } = asset.data;
+  delete asset.data.caption.content;
```
Why do we need to delete this?

ikovac replied:
The `caption.content` property is deleted before saving to the database. If the property is kept, it will be stored in the database but I don't think it's necessary.

MiroDojkic commented on src/server/sproutVideo/caption.js:
```diff
@@ -0,0 +1,17 @@
+  }
+}
+
+module.exports = Captions;
```
Does it make sense to use singular nouns for these clients (also to align them with the module names)?

ikovac committed:
Refactor toolbar component

ikovac committed:
Refactor is captionUploadBtn disabled logic

ikovac committed:
Change Captions and Videos classes to singular nouns

ikovac committed:
Merge branch 'feature/implement-sprout-video' into feature/captions

ikovac committed:
Make on save handler more readable

Use lodash omit and get methods

ikovac committed:
Merge branch 'feature/captions' of https://github.com/ExtensionEngine/tce-sprout-video into feature/captions

# Conflicts:
#	dist/tce-sprout-video.umd.min.js

ikovac committed:
Ensure nonpersistent props are deleted before save

MiroDojkic commented on src/edit/CaptionUpload.vue:
```diff
@@ -0,0 +1,84 @@
+            v-on="on"
+            v-bind="attrs"
+            class="delete-caption"
+            color="error">
```
I don't think we're using `error` color for this icon anywhere in the app?
There's also no reason to make it more important then other things in the toolbar so we should probably stick to a more neutral color.

MiroDojkic commented on src/edit/Toolbar.vue:
```diff
@@ -1,60 +1,34 @@
-    upload(e) {
-      const [file] = e.target.files;
-      this.$elementBus.emit('save', { file });
+    onSave(payload) {
```
Should we rename this to `emitSave` to make it simpler to figure out what's going from reading the template?

MiroDojkic commented on src/edit/UploadBtn.vue:
```diff
@@ -28,3 +28,9 @@ export default {
+
+<style lang="scss" scoped>
+.upload-btn {
+  height: 100% !important;
```
Can we beat the specificity here with something like `.upload-btn.v-btn` or similar?

MiroDojkic commented on src/server/index.js:
```diff
@@ -4,47 +4,88 @@ const { createClient } = require('./sproutVideo');
+  const { id: videoId, playable } = asset.data.video;
+  if (!videoId || !playable) return asset;
+  await processCaption(asset, client);
+  deleteNonPersistentAssetProps(asset);
```
`NonPersistent` -> `Temporary`? 😄

ikovac committed:
💄Change naming

ikovac committed:
Remove !important inside upload btn css

ikovac committed:
Inline attributes

ikovac committed:
Delete token and uploadUrl before saving to the db

MiroDojkic commented:
@ikovac please resolve conflicts.

ikovac committed:
Merge base branch 🔀

ikovac committed:
🚀

# Conflicts:
#	dist/server/index.js
#	dist/tce-sprout-video.css
#	dist/tce-sprout-video.es.js
#	dist/tce-sprout-video.js
#	dist/tce-sprout-video.umd.js
#	dist/tce-sprout-video.umd.min.js
#	dist/tce-sprout-video.umd.min.js.map
#	src/edit/index.vue
#	src/server/index.js

ikovac commented:
@markolovric @kjuej This PR should work properly after merging:
https://github.com/ExtensionEngine/tailor/pull/752

kjuej commented:
@ikovac @MiroDojkic Seems to be working as expected.
